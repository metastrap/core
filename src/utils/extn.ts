// eslint-disable-next-line import/prefer-default-export
export function getExtn(filename: string) {
  return /\.(\w+)$/.exec(filename)?.[1];
}
