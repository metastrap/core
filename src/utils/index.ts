export * as astUtils from './ast';
export * as logger from './logger';
export * as read from './read';
