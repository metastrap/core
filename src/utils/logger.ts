/* eslint-disable no-console */
export default function logger(...args: unknown[]) {
  console.log(...args);
}
