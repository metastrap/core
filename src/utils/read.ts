import JSZip from 'jszip';

export default async function readZipToMap(
  zip: JSZip,
): Promise<Map<string, string>> {
  const files = Object.keys(zip.files);

  const maybeContentList = files.map(async (file) => {
    if (!zip.files[file].dir) {
      const contents = await zip.files[file].async('text');
      return [
        file,
        contents || '',
      ] as [string, string];
    }
    /* directories will return null */
    return null;
  });

  const contentList = (
    await Promise.all(maybeContentList)
  ).filter((file) => !!file);

  return new Map(contentList);
}
