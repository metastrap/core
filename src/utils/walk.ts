/* https://gist.github.com/lovasoa/8691344?permalink_comment_id=2927279#gistcomment-2927279 */
import { promises as fs } from 'fs';
import path from 'path';

export async function walk(dir: string) {
  const fileList = await fs.readdir(dir);
  const files: string[] = await Promise.all(
    fileList.map(async (file) => {
      const filePath = path.join(dir, file);
      const stats = await fs.stat(filePath);
      if (stats.isDirectory()) return walk(filePath);
      if (stats.isFile()) return filePath;
      return null;
    }).filter((file) => !!file) as Promise<string>[],
  );

  return files.reduce(
    (all, folderContents) => all.concat(folderContents || ''),
    [] as unknown as string[],
  );
}

export async function readFilesToMap(
  dir: string,
): Promise<Map<string, string>> {
  const files = await walk(dir);

  return new Map(
    await Promise.all(
      files.map(async (file) => {
        const contents = await fs.readFile(file, 'utf8');
        return [
          file.replace(`${dir}/`, ''),
          contents || '',
        ] as [string, string];
      }),
    ),
  );
}
