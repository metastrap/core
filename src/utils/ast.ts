/* browser */
import { ParseResult, ParserPlugin, parse } from '@babel/parser';
import generate from '@babel/generator';

import type { File } from '@babel/types';

import { TAstMap, TAstValue } from 'types';
import logger from './logger';

enum EFileFeature {
  JAVASCRIPT = 1,
  TYPESCRIPT = 2,
  JSX = 4,
  JSON = 8,
}

export function convertTextToAst(
  fileContentMap: Map<string, string>,
): TAstMap {
  const resultMap = new Map<string, TAstValue>();

  fileContentMap.forEach((content, file) => {
    let featureFlag = 0;
    const extname = file.split('.').pop() || '';
    const babelParsePlugins: ParserPlugin[] = ['classProperties'];

    let result: TAstValue = content;
    switch (extname) {
      case 'js':
        featureFlag |= EFileFeature.JAVASCRIPT;
        break;
      case 'ts':
        featureFlag |= EFileFeature.TYPESCRIPT;
        break;
      case 'jsx':
        featureFlag |= EFileFeature.JSX;
        break;
      case 'tsx':
        featureFlag |= EFileFeature.TYPESCRIPT | EFileFeature.JSX;
        break;
      case 'json':
        try {
          result = JSON.parse(content);
        } catch (e) {
          logger((e as Error).message);
        }
    }

    if (featureFlag & EFileFeature.TYPESCRIPT) babelParsePlugins.push('typescript');
    if (featureFlag & EFileFeature.JSX) babelParsePlugins.push('jsx');

    if (featureFlag & (EFileFeature.JAVASCRIPT | EFileFeature.TYPESCRIPT | EFileFeature.JSX)) {
      try {
        result = parse(content, {
          sourceType: 'module',
          plugins: babelParsePlugins,
        });
      } catch (e) {
        logger((e as Error).message);
        result = content;
      }
    }
    resultMap.set(file, result);
  });
  return resultMap;
}

export function print(ast: Map<string, TAstValue>): Map<string, string> {
  const resultMap = new Map<string, string>();
  ast.forEach((astValue, file) => {
    if (typeof astValue === 'string') {
      resultMap.set(file, astValue);
      return;
    }
    if (astValue.program) {
      resultMap.set(file, generate(astValue as ParseResult<File>).code);
      return;
    }
    if (typeof astValue === 'object') {
      try {
        resultMap.set(file, JSON.stringify(astValue, null, 2));
      } catch (e) {
        resultMap.set(file, astValue.toString?.() || '');
      }
    }
  });
  return resultMap;
}
