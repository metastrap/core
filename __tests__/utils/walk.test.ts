import path from 'node:path';
import { walk, readFilesToMap } from '@/utils/walk';

describe('utils/walk', () => {
  test('walk through fixtures', async () => {
    const fixtures = await walk(path.join(process.cwd(), 'fixtures', 'next'));
    expect(fixtures.map(
      (file) => file.replace(`${process.cwd()}/fixtures/next/`, ''),
    )).toMatchSnapshot();
  });

  test('read files to map', async () => {
    const fixtureMap = await readFilesToMap(path.join(process.cwd(), 'fixtures', 'next'));
    expect(fixtureMap).toMatchSnapshot();
  });
});
