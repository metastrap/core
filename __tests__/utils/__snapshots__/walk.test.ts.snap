// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/walk read files to map 1`] = `
Map {
  ".gitignore" => "# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.

# dependencies
/node_modules
/.pnp
.pnp.js
.yarn/install-state.gz

# testing
/coverage

# next.js
/.next/
/out/

# production
/build

# misc
.DS_Store
*.pem

# debug
npm-debug.log*
yarn-debug.log*
yarn-error.log*

# local env files
.env*.local

# vercel
.vercel

# typescript
*.tsbuildinfo
next-env.d.ts
",
  "app/globals.css" => "@tailwind base;
@tailwind components;
@tailwind utilities;
",
  "app/layout.jsx" => "export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body>{children}</body>
    </html>
  );
}
",
  "app/page.jsx" => "export default function Home() {
  return (
    <p>hello</p>
  );
}
",
  "jsconfig.json" => "{
  "compilerOptions": {
    "paths": {
      "@/*": ["./*"]
    }
  }
}
",
  "next.config.js" => "/** @type {import('next').NextConfig} */
const nextConfig = {};

module.exports = nextConfig;
",
  "package.json" => "{
  "name": "hello",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "react": "^18",
    "react-dom": "^18",
    "next": "14.0.4"
  }
}
",
  "postcss.config.js" => "module.exports = {
  plugins: {
    tailwindcss: {},
    autoprefixer: {},
  },
};
",
  "tailwind.config.js" => "/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [
    './pages/**/*.{js,ts,jsx,tsx,mdx}',
    './components/**/*.{js,ts,jsx,tsx,mdx}',
    './app/**/*.{js,ts,jsx,tsx,mdx}',
  ],
  plugins: [],
};
",
}
`;

exports[`utils/walk walk through fixtures 1`] = `
[
  ".gitignore",
  "app/globals.css",
  "app/layout.jsx",
  "app/page.jsx",
  "jsconfig.json",
  "next.config.js",
  "package.json",
  "postcss.config.js",
  "tailwind.config.js",
]
`;
